### PR TITLE
Fix 29932 - sync put!/take! for unbuffered RemoteChannels

### LINKED ISSUE #29932: put! and take! Operations on Remote Channels not atomic
I have been implementing a parallel version of the [Reduce kernel](https://github.com/ParRes/Kernels), implementing a tree reduce method. After processes calculate their partial sums (partial sums in this instance are vectors), they `put!` to a blocking channel to avoid overwriting the data before they communicate it to other processes. For sufficiently large vectors, it is possible for the data to be mutated before it is serialised.

This behaviour is illustrated by the following minimal example. Worker two will take a zero vector of sufficient size and add one to it at each iteration. Immediately afterwards, it will initiate a blocking communication with worker three.  Worker three will assert that the data contained within the vector at each step is equal to the iteration counter.

`put!` and `take!` operations appear to atomically serialise the vector before beginning the next iteration for smaller vectors, however this semantic is not maintained for larger vectors.

```
@fetchfrom 2 begin
    global inbox = RemoteChannel(()->Channel{Vector{Float64}}(0))
end

@fetchfrom 3 begin
    global outbox = @fetchfrom 2 inbox
end

@sync begin
    @spawnat 2 begin
        vector = zeros(8200)
        for i in 1 : 1000
            vector .+= 1.0
            put!(inbox,vector)
        end
    end

    @spawnat 3 begin
        local incoming
        output = [take!(outbox) for i in 1 : 1000]
        wrong = 0
        for (i,v) in enumerate(output)
            if v[1] != i
                wrong += 1
            end
        end
        println("$wrong failed")
    end
end
```

```
Julia Version 1.0.1
Commit 0d713926f8 (2018-09-29 19:05 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, sandybridge)
```

### PR DESCRIPTION
**Pull request by amitmurthy**
This PR is in 3 commits and closes JuliaLang/julia#29932 .

1) An argname `callee` has been corrected to `caller` which it should have been all along. No change in logic. 

2) Fixes JuliaLang/julia#29932 for the specific case of local `put!` with a remote `take!` on an unbuffered `RemoteChannel`.  Does not affect any other use case. It required the `put!` call to wait till the locally proxied `take!` to be serialized completely before returning. Locking / unlocking was outside lexical scope in the `take!` scenario and hence the slightly convoluted solution.

3) Documents behavior of local invocations of remote calls in the manual. Differences w.r.t. arguments in these two cases are highlighted.

### CODE REVIEW

**Review by KristofferC — COMMENTED** [review]
Perhaps use `jldoctest` instead of `julia-repl` so that these are tested in the doctests?

**Comment by amitmurthy** [comment]
@KristofferC does `jldoctest` check for exact matches on each line? If so, some `objectid`s currently being printed in the examples will need to be replaced...

**Comment by KristofferC** [comment]
Yes, but it is possible to use regex filtering to account for this https://juliadocs.github.io/Documenter.jl/latest/man/doctests/#Filtering-Doctests-1.

Here is an example of it used in Base: https://github.com/JuliaLang/julia/blame/master/doc/src/manual/performance-tips.md#L97.

**Comment by rohanmclure** [comment]
Do we intend to synchronise `put!` and `take!` operations for _remote_ communication, as in #29932?

It does seem that there is still no way to obtain synchronisation by means of channels for sufficiently large arrays.

**Comment by JeffBezanson** [comment]
I don't think this should be called `size`, since that has a well-established meaning as the tuple of sizes of an n-d array. Maybe `capacity`, or `maxlength`?

File: stdlib/Distributed/src/process_messages.jl [hunk 1]
```diff
@@ -10,11 +10,35 @@ mutable struct RemoteValue
...
+    synctake::Union{ReentrantLock, Nothing}
+
+    function RemoteValue(c)
+        sz = -1
+        try
+            sz = size(c)
+        catch
+        end
+
+        if sz == 0
```
Comment (JeffBezanson) [inline stdlib/Distributed/src/process_messages.jl]: Doesn't `isbuffered` do this?

Comment (amitmurthy) [reply stdlib/Distributed/src/process_messages.jl]: You are right. Will update the PR.

**Comment by amitmurthy** [comment]
PR updated - Imports `isbuffered(::Channel)` from Base and uses `jldoctest`.

**Comment by amitmurthy** [comment]
@JeffBezanson , the OSX Travis failure appears unrelated. Please have another look and merge if OK.

On unbuffered `Channel`s `put!` blocks till another task performs a `take!` and vice-versa. Hence it is *reasonable expectation* that unbuffered `RemoteChannel`s too behave the same way. However, since `put!` on a buffered channel is not expected to block as long there is spare capacity in the channel, we do not synchronise put/take for buffered remote channels. The manual section highlights the fact that `put!` on a buffered, locally owned  `RemoteChannel`s stores a reference to the object being put (not a copy),  and any issues that may arise due to this.

File: stdlib/Distributed/src/remotecall.jl [hunk 2]
```diff
@@ -549,18 +549,28 @@ function put!(rr::Future, v)
...
 
 
 put!(rv::RemoteValue, args...) = put!(rv.c, args...)
-put_ref(rid, args...) = (put!(lookup_ref(rid), args...); nothing)
+function put_ref(rid, caller, args...)
+    rv = lookup_ref(rid)
+    put!(rv, args...)
+    if myid() == caller && rv.synctake !== nothing
+        # Wait till a "taken" value is serialized out - github issue #29932
+        yield()
```
Comment (JeffBezanson) [inline stdlib/Distributed/src/remotecall.jl]: Is this `yield` actually necessary? It seems to me the `lock` is enough.

Comment (amitmurthy) [reply stdlib/Distributed/src/remotecall.jl]: Not required. Have removed it and updated the PR.

**Pull request merged**